gdcdatamodel/models/redaction.py

```diff
@@ -0,0 +1,129 @@
+import copy
+
+from datetime import datetime
+
+from sqlalchemy import Boolean, Column, DateTime, Enum, ForeignKey, Integer, String, Text, text
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.ext.hybrid import hybrid_property
+from sqlalchemy.orm import relationship
+
+Base = declarative_base()
+
+
+class RedactionLog(Base):
+    """
+    Logs a redaction event, each redacted node will be stored as a RedactionEntry
+    """
+
+    __tablename__ = 'redaction_log'
+
+    id = Column(Integer, primary_key=True, nullable=False)
+    annotation_id = Column(String(64), nullable=False, unique=True)
+
+    # who initiated the redaction
+    initiated_by = Column(String(64), nullable=False, index=True)
+
+    # who rescinded this redaction
+    rescinded_by = Column(String(64), nullable=True)
+
+    # reasons for redaction
+    reason = Column(Text, nullable=False)  # long text
+    reason_category = Column(String(128), nullable=False, index=True)  # short desc
+
+    project_id = Column(String(32), nullable=False, index=True)
+
+    # LATEST => only latest version is redacted
+    # COMPLETE => all versions are redacted
+    # PREVIOUS => All previous versions are redacted
+    # applicable only to files on indexd
+    redaction_type = Column(Enum("LATEST", "COMPLETE", "PREVIOUS", name="redaction_types"), default="COMPLETE")
+
+    created_datetime = Column(
+        DateTime(timezone=True),
+        nullable=False,
+        server_default=text('now()'),
+    )
+
+    date_rescinded = Column(
+        DateTime(timezone=True),
+        nullable=True
+    )
+
+    entries = relationship("RedactionEntry", back_populates="redaction_log")  # type: list[RedactionEntry]
+
+    @hybrid_property
+    def project(self):
+        return self.project_id.split("-")[1]
+
+    @hybrid_property
+    def program(self):
+        return self.project_id.split("-")[0]
+
+    def rescind_all(self, rescinded_by):
+        """Rescinds all entries on this redaction log"""
+        self.rescinded_by = rescinded_by
+        self.date_rescinded = datetime.now()
+        for entry in self.entries:
+            entry.rescind(rescinded_by)
+
+    @hybrid_property
+    def is_rescinded(self):
+        """Checks if all redacted entries in this log has been rescinded
+        Returns:
+           bool: True if all are rescinded
+        """
+        for entry in self.entries:
+            if not entry.rescinded:
+                return False
+        return True
+
+    def to_json(self):
+        json = dict(id=self.id, annotation_id=self.annotation_id,
+                    project_id=self.project_id, initiated_by=self.initiated_by, reason=self.reason)
+        return json
+
+
+class RedactionEntry(Base):
+    """
+    Logs a redacted node, holds enough information to enable
```

Comment:
to enable what?

Reply:
:trollface:

Reply:
it just felt incomplete and I couldn't fall asleep :trollface: